crates/subspace-farmer-components/src/plotting.rs

```diff
@@ -151,34 +151,75 @@ pub enum PlottingError {
     },
 }
 
-/// Plot a single sector, where `sector` and `sector_metadata` must be positioned correctly at the
-/// beginning of the sector (seek to desired offset before calling this function and seek back
-/// afterwards if necessary).
+/// Options for plotting a sector.
 ///
 /// Sector output and sector metadata output should be either empty (in which case they'll be
 /// resized to correct size automatically) or correctly sized from the beginning or else error will
 /// be returned.
+pub struct PlotSectorOptions<'a, PosTable, PG>
+where
+    PosTable: Table,
+{
+    /// Public key corresponding to sector
+    pub public_key: &'a PublicKey,
+    /// Sector index
+    pub sector_index: SectorIndex,
+    /// Getter for pieces of archival history
+    pub piece_getter: &'a PG,
+    /// Retry policy for piece getter
+    pub piece_getter_retry_policy: PieceGetterRetryPolicy,
+    /// Farmer protocol info
+    pub farmer_protocol_info: &'a FarmerProtocolInfo,
+    /// KZG instance
+    pub kzg: &'a Kzg,
+    /// Erasure coding instance
+    pub erasure_coding: &'a ErasureCoding,
+    /// How many pieces should sector contain
+    pub pieces_in_sector: u16,
+    /// Where plotted sector should be written, vector must either be empty (in which case it'll be
+    /// resized to correct size automatically) or correctly sized from the beginning
+    pub sector_output: &'a mut Vec<u8>,
+    /// Where plotted sector metadata should be written, vector must either be empty (in which case
+    /// it'll be resized to correct size automatically) or correctly sized from the beginning
+    pub sector_metadata_output: &'a mut Vec<u8>,
+    /// Semaphore for part of the plotting when farmer downloads new sector, allows to limit memory
+    /// usage of the plotting process, permit will be held until the end of the plotting process,
+    /// so must always be larger than `encoding_semaphore` or else plotting may cause deadlock
```

Comment:
Do you mind explaining the call sequence leading to deadlock when d_sem capacity is less than e_sem capacity? At first glance, it should work because we acquire semaphores in the same order every time and for a limited amount of time (d_sem first then e_sem). What am I missing? 

Anyway, if it's indeed the case that the deadlock is possible due to the configuration issues - we should introduce a check and fail much earlier than during the plotting.

Reply:
Hm, I think you're right, it'll reduce performance, but there should be no deadlock, will update comments.